File: commands/qrun/daapi_c/app/src/run_primitive.c
/*
# This code is part of Qiskit.
#
# (C) Copyright IBM 2025.
#
# This code is licensed under the Apache License, Version 2.0. You may
# obtain a copy of this license in the LICENSE.txt file in the root directory
# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
#
# Any modifications or derivative works of this code must retain this
# copyright notice, and modified files need to carry a notice indicating
# that they have been altered from the originals.
*/
#include <stdint.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>

#include "cjson/cJSON.h"
#include "direct_access_capi.h"

extern const char *IAM_APIKEY;
extern const char *SERVICE_CRN;
extern const char *IAM_ENDPOINT;
extern const char *S3_ENDPOINT;
extern const char *AWS_ACCESS_KEY_ID;
extern const char *AWS_SECRET_ACCESS_KEY;
extern const char *S3_REGION;
extern const char *S3_BUCKET;
extern const char *DAAPI_ENDPOINT;

int main(int argc, char *argv[]) {

  int rc = 0;

  if (argc != 4) {
    printf("run_primitive <backend> <primitive type> <PUBs JSON file>\n");
    return -1;
  }

  ProgramId type;
  if (strcmp(argv[2], "sampler") == 0)
    type = SAMPLER;
  else if (strcmp(argv[2], "estimator") == 0)
    type = ESTIMATOR;
  else {
    printf("Unknown primitive type: %s\n", argv[2]);
    return -1;
  }

  struct ClientBuilder *builder = daapi_bldr_new(DAAPI_ENDPOINT);
  if (!builder) {
    printf("Failed to create a builder.\n");
    return -1;
  }

  rc = daapi_bldr_enable_iam_auth(builder, IAM_APIKEY, SERVICE_CRN,
                                  IAM_ENDPOINT);
  if (rc < 0)
    printf("Failed to enable IAM auth. rc=%d\n", rc);

  rc = daapi_bldr_set_timeout(builder, 60.0);
  if (rc < 0)
    printf("Failed to enable timeout. rc=%d\n", rc);

  rc = daapi_bldr_set_exponential_backoff_retry(builder, 5, 2, 1, 10);
  if (rc < 0)
    printf("Failed to enable retries. rc=%d\n", rc);

  rc = daapi_bldr_set_s3_bucket(builder, AWS_ACCESS_KEY_ID,
                                AWS_SECRET_ACCESS_KEY, S3_ENDPOINT, S3_BUCKET,
                                S3_REGION);
  if (rc < 0)
    printf("Failed to set S3 params. rc=%d\n", rc);

  struct Client *client = daapi_cli_new(builder);
  if (!client) {
    daapi_free_builder(builder);
    return -1;
  }
  rc = daapi_free_builder(builder);

  FILE *fp = fopen(argv[3], "r");
  if (!fp) {
    printf("Failed to open PUBs file (%s)\n", argv[1]);
    return -1;
  }

  fseeko(fp, 0, SEEK_END);
  long size = ftello(fp);
  char *bufp = malloc(size);
  char *curr_ptr = bufp;
  fseeko(fp, 0, SEEK_SET);
  while(size > 0) {
    size_t sz = fread(curr_ptr, 1, size, fp);
    size -= sz;
    curr_ptr += sz;
  }
  fclose(fp);

  struct PrimitiveJob *job = daapi_cli_run_primitive(
      client, argv[1], type, 300, DEBUG, bufp, NULL);
  free(bufp);
  if (job) {
    JobStatus final_state;
    bool is_running = false;
    bool is_in_final_state = false;

    rc = daapi_prim_is_running(job, &is_running);
    printf("is_running %d, %d\n", rc, is_running);
    rc = daapi_prim_is_in_final_state(job, &is_in_final_state);
    printf("is_in_final_state %d, %d\n", rc, is_in_final_state);

    const char *job_id = daapi_prim_get_job_id(job);
    printf("JOB ID = %s\n", job_id);

    rc = daapi_prim_wait_for_final_state(job, &final_state);
    printf("wait_for rc=%d, final_state=%d\n", rc, final_state);

    const char *result = daapi_prim_get_result_as_string(job);
    if (result) {
      cJSON *json = cJSON_Parse(result);
      printf("%s\n", cJSON_PrintUnformatted(json));
      cJSON_free(json);
      daapi_free_string((char *)result);
    }

    const char *logs = daapi_prim_get_logs(job);
    if (logs) {
      printf("%s\n", logs);
      daapi_free_string((char *)logs);
    }

    struct Metrics* metrics = daapi_cli_get_metrics(client, job_id);
    if (metrics) {
      printf("created_time: %s, end_time: %s, quantum_nanoseconds: %ld\n",
             metrics->created_time, metrics->end_time,
             metrics->quantum_nanoseconds);
      daapi_free_metrics(metrics);
    }

    rc = daapi_free_primitive(job);
    if (rc < 0) {
      printf("Failed to free Job(%p). rc=%d\n", job, rc);
    }

    rc = daapi_cli_delete_job(client, job_id);
    if (rc < 0) {
      printf("Failed to delete Job(%s). rc=%d\n", job_id, rc);
    }

    rc = daapi_free_string((char *)job_id);
    if (rc < 0) {
      printf("Failed to free job ID. rc=%d\n", rc);
    }
  }

  rc = daapi_free_client(client);
  if (rc < 0)
    printf("Failed to free Client(%p). rc=%d\n", client, rc);

  return 0;
}

File: qrmi/dependencies/direct_access_client/src/api/run_primitive.rs
//
// (C) Copyright IBM 2024, 2025
//
// This code is licensed under the Apache License, Version 2.0. You may
// obtain a copy of this license in the LICENSE.txt file in the root directory
// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
//
// Any modifications or derivative works of this code must retain this
// copyright notice, and modified files need to carry a notice indicating
// that they have been altered from the originals.

use crate::Client;
use anyhow::{bail, Context, Result};
use log::debug;
use uuid::Uuid;

use aws_sdk_s3::error::DisplayErrorContext;
use serde::de::DeserializeOwned;

use crate::models::jobs::{JobStatus, LogLevel, ProgramId};
use crate::PrimitiveJob;

const S3KEY_INPUT_PREFIX: &str = "input_";
const S3KEY_RESULTS_PREFIX: &str = "results_";
const S3KEY_LOGS_PREFIX: &str = "logs_";

impl Client {
    /// Invokes a Qiskit Runtime primitive. Parameters to inject into the primitive are defined in [EstimatorV2 input](https://github.com/Qiskit/ibm-quantum-schemas/blob/main/schemas/estimator_v2_schema.json) and [SamplerV2 input](https://github.com/Qiskit/ibm-quantum-schemas/blob/main/schemas/sampler_v2_schema.json).
    /// [`Client`] needs to be created by the [`ClientBuilder`](crate::ClientBuilder) with [`with_s3bucket`](crate::ClientBuilder::with_s3bucket) to use this function.
    ///
    /// # Example
    ///
    /// ```no_run
    /// #[tokio::main]
    /// async fn main() -> Result<(), Box<dyn std::error::Error>> {
    ///     use serde_json::json;
    ///     use direct_access_api::{AuthMethod, ClientBuilder, models::ProgramId, models::LogLevel};
    ///
    ///     let payload = json!({
    ///         "pubs":[
    ///             [
    ///                 "OPENQASM 3.0; include \\\"stdgates.inc\\\"; bit[2] meas; rz(pi/2) $0; sx $0; rz(pi/2) $0; cx $0, $1; meas[0] = measure $0; meas[1] = measure $1;",[],128
    ///             ],
    ///         ],
    ///         "supports_qiskit": false,
    ///         "version":2,
    ///     });
    ///
    ///     let client = ClientBuilder::new("http://localhost:8290")
    ///         .with_auth(AuthMethod::IbmCloudIam {
    ///             apikey: "your_iam_apikey".to_string(),
    ///             service_crn: "your_service_crn".to_string(),
    ///             iam_endpoint_url: "iam_endpoint_url".to_string(),
    ///         })
    ///         .with_s3bucket(
    ///             "my_aws_access_key_id",
    ///             "my_aws_secret_access_key",
    ///             "http://localhost:9000",
    ///             "my_bucket",
    ///             "us-east-1",
    ///         )
    ///         .build()
    ///         .unwrap();
    ///
    ///     let _primitive_job = client
    ///         .run_primitive("ibm_brisbane", ProgramId::Sampler, 3600, LogLevel::Info, &payload, None)
    ///         .await?;
    ///     Ok(())
    /// }
    /// ```
    ///
    /// # Errors
    ///
    /// This function will return an error variant when:
    /// - connection failed.
    /// - invalid input is received.
    /// - authentication failed.
    /// - the request took too long to read.
    /// - a duplicate job with same job ID was submitted.
    /// - the request body is too large.
    /// - validation of the request failed. The error message contains details about the specific validation error.
    /// - backend is reserved and jobs outside of the reservation cannot be run.
    /// - per backend concurrent job limit has been reached.
    /// - S3 authentication failed.
    /// - S3 connection failed.
    /// - S3 bucket is not found.
    ///
    pub async fn run_primitive(
        &self,
        backend: &str,
        program_id: ProgramId,
        timeout_secs: u64,
        log_level: LogLevel,
        payload: &serde_json::Value,
        job_id: Option<String>,
    ) -> Result<PrimitiveJob> {
        let s3_config = self.s3_config.clone().context(
            "S3 bucket is not configured. Use ClientBuilder.with_s3_bucket() to use this function.",
        )?;

        let s3_client = aws_sdk_s3::Client::from_conf(s3_config);
        let id;
        if let Some(value) = job_id {
            id = value;
        } else {
            id = Uuid::new_v4().to_string();
        }
        let s3_bucket = self.s3_bucket.clone().unwrap();

        let converted_vec = serde_json::to_vec::<serde_json::Value>(payload)?;
        let job_param_key = format!("{}{}.json", S3KEY_INPUT_PREFIX, id);
        let _ = match s3_client
            .put_object()
            .bucket(s3_bucket.clone())
            .key(job_param_key.clone())
            .body(converted_vec.into())
            .send()
            .await
        {
            Ok(val) => val,
            Err(err) => {
                bail!(format!(
                    "An error occurred during upload to S3: {}",
                    DisplayErrorContext(&err)
                ));
            }
        };

        let input_presigned_url =
            crate::storages::s3::get_presigned_url(&s3_client, &s3_bucket, &job_param_key).await?;

        let results_key = format!("{}{}.json", S3KEY_RESULTS_PREFIX, id);
        let results_presigned_url =
            crate::storages::s3::get_presigned_url_for_put(&s3_client, &s3_bucket, &results_key)
                .await?;

        let logs_key = format!("{}{}.json", S3KEY_LOGS_PREFIX, id);
        let logs_presigned_url =
            crate::storages::s3::get_presigned_url_for_put(&s3_client, &s3_bucket, &logs_key)
                .await?;

        let job_param = serde_json::json!({
            "id": id,
            "backend": backend.to_string(),
            "program_id": program_id.to_string(),
            "log_level": log_level.to_string(),
            "timeout_secs": timeout_secs,
            "storage": {
                "input": {
                    "presigned_url": input_presigned_url,
                    "type": "s3_compatible".to_string(),
                },
                "results": {
                    "presigned_url": results_presigned_url,
                    "type": "s3_compatible".to_string(),
                },
                "logs": {
                    "presigned_url": logs_presigned_url,
                    "type": "s3_compatible".to_string(),
                },
            }
        });
        let job_id = self.run_job(&job_param).await?;
        Ok(PrimitiveJob {
            job_id,
            client: self.clone(),
            s3_client,
            s3_bucket,
        })
    }
}

impl PrimitiveJob {
    /// Return the results of the job.
    ///
    /// # Example
    ///
    /// ```no_run
    /// #[tokio::main]
    /// async fn main() -> Result<(), Box<dyn std::error::Error>> {
    ///     use serde_json::json;
    ///     use direct_access_api::{AuthMethod, ClientBuilder, models::ProgramId, models::LogLevel};
    ///
    ///     let payload = json!({
    ///         "pubs":[
    ///             [
    ///                 "OPENQASM 3.0; include \\\"stdgates.inc\\\"; bit[2] meas; rz(pi/2) $0; sx $0; rz(pi/2) $0; cx $0, $1; meas[0] = measure $0; meas[1] = measure $1;",[],128
    ///             ],
    ///         ],
    ///         "supports_qiskit": false,
    ///         "version":2,
    ///     });
    ///
    ///     let client = ClientBuilder::new("http://localhost:8290")
    ///         .with_auth(AuthMethod::IbmCloudIam {
    ///             apikey: "your_iam_apikey".to_string(),
    ///             service_crn: "your_service_crn".to_string(),
    ///             iam_endpoint_url: "iam_endpoint_url".to_string(),
    ///         })
    ///         .with_s3bucket(
    ///             "my_aws_access_key_id",
    ///             "my_aws_secret_access_key",
    ///             "http://localhost:9000",
    ///             "my_bucket",
    ///             "us-east-1",
    ///         )
    ///         .build()
    ///         .unwrap();
    ///
    ///     let primitive_job = client
    ///         .run_primitive("ibm_brisbane", ProgramId::Sampler, 3600, LogLevel::Info, &payload, None)
    ///         .await?;
    ///     let _result = primitive_job.get_result::<serde_json::Value>().await?;
    ///     Ok(())
    /// }
    /// ```
    ///
    /// # Errors
    ///
    /// This function will return an error variant when:
    /// - Result is not available since job was not succeeded.
    /// - Result is not available until the job is completed.
    /// - S3 authentication failed.
    /// - S3 connection failed.
    /// - S3 bucket is not found.
    ///
    pub async fn get_result<T: DeserializeOwned>(&self) -> Result<T> {
        let status = self.get_status().await?;
        if JobStatus::Cancelled == status || JobStatus::Failed == status {
            bail!("Result is not available since job was not succeeded.".to_string());
        } else if JobStatus::Running == status {
            bail!("Result is not available until the job is completed.".to_string());
        }

        let key = format!("{}{}.json", S3KEY_RESULTS_PREFIX, self.job_id);
        let presigned_url =
            crate::storages::s3::get_presigned_url(&self.s3_client, &self.s3_bucket, &key).await?;
        debug!("{}", presigned_url);

        let client = reqwest::Client::new();
        let resp = client
            .get(presigned_url)
            .header("Content-Type", "application/json")
            .send()
            .await?;
        if resp.status().is_success() {
            let json_data = resp.json::<T>().await?;
            Ok(json_data)
        } else {
            let json_data = resp.json::<serde_json::Value>().await?;
            bail!(format!("{:?}", json_data))
        }
    }

    /// Return the logs of the job.
    ///
    /// # Example
    ///
    /// ```no_run
    /// #[tokio::main]
    /// async fn main() -> Result<(), Box<dyn std::error::Error>> {
    ///     use serde_json::json;
    ///     use direct_access_api::{AuthMethod, ClientBuilder, models::ProgramId, models::LogLevel};
    ///
    ///     let payload = json!({
    ///         "pubs":[
    ///             [
    ///                 "OPENQASM 3.0; include \\\"stdgates.inc\\\"; bit[2] meas; rz(pi/2) $0; sx $0; rz(pi/2) $0; cx $0, $1; meas[0] = measure $0; meas[1] = measure $1;",[],128
    ///             ],
    ///         ],
    ///         "supports_qiskit": false,
    ///         "version":2,
    ///     });
    ///
    ///     let client = ClientBuilder::new("http://localhost:8290")
    ///         .with_auth(AuthMethod::IbmCloudIam {
    ///             apikey: "your_iam_apikey".to_string(),
    ///             service_crn: "your_service_crn".to_string(),
    ///             iam_endpoint_url: "iam_endpoint_url".to_string(),
    ///         })
    ///         .with_s3bucket(
    ///             "my_aws_access_key_id",
    ///             "my_aws_secret_access_key",
    ///             "http://localhost:9000",
    ///             "my_bucket",
    ///             "us-east-1",
    ///         )
    ///         .build()
    ///         .unwrap();
    ///
    ///     let primitive_job = client
    ///         .run_primitive("ibm_brisbane", ProgramId::Sampler, 3600, LogLevel::Info, &payload, None)
    ///         .await?;
    ///     let _logs = primitive_job.get_logs().await?;
    ///     Ok(())
    /// }
    /// ```
    ///
    /// # Errors
    ///
    /// This function will return an error variant when:
    /// - Logs are not available until the job is in its final state.
    /// - S3 authentication failed.
    /// - S3 connection failed.
    /// - S3 bucket is not found.
    ///
    pub async fn get_logs(&self) -> Result<String> {
        let in_final_state = self.is_in_final_state().await?;
        if !in_final_state {
            bail!("Logs are not available until the job is in its final state.".to_string());
        }

        let key = format!("{}{}.json", S3KEY_LOGS_PREFIX, self.job_id);
        let presigned_url =
            crate::storages::s3::get_presigned_url(&self.s3_client, &self.s3_bucket, &key).await?;
        debug!("{}", presigned_url);

        let client = reqwest::Client::new();
        let resp = client
            .get(presigned_url)
            .header("Content-Type", "application/json")
            .send()
            .await?;
        let status = resp.status();
        let text_data = resp.text().await?;
        if status.is_success() {
            Ok(text_data)
        } else {
            bail!(text_data)
        }
    }
}
